core/src/androidTest/java/com/pierfrancescosoffritti/androidyoutubeplayer/core/YouTubePlayerViewClickTest.kt

```diff
@@ -0,0 +1,116 @@
+package com.pierfrancescosoffritti.androidyoutubeplayer.core
+
+import android.app.Activity
+import android.view.MotionEvent
+import android.view.View
+import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.action.ViewActions.*
+import androidx.test.espresso.matcher.ViewMatchers.withId
+import androidx.test.filters.LargeTest
+import androidx.test.rule.ActivityTestRule
+import com.pierfrancescosoffritti.androidyoutubeplayer.core.player.views.YouTubePlayerView
+import com.pierfrancescosoffritti.androidyoutubeplayer.core.testActivity.TestActivity
+import com.pierfrancescosoffritti.androidyoutubeplayer.test.R
+import org.junit.Assert.*
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+import org.mockito.Mockito.mock
+import org.mockito.Mockito.verify
+import java.util.*
+
+@LargeTest
+class YouTubePlayerViewClickTest {
+
+    @Rule
+    @JvmField
+    val activityRule = ActivityTestRule(TestActivity::class.java)
+    var testActivity: Activity? = null
+
+    var youTubePlayerView: YouTubePlayerView? = null
+
+    val clickTestViewId = R.id.youtube_player_view
+
+    @Before
+    fun setup() {
+        testActivity = activityRule.getActivity()
+        youTubePlayerView = testActivity!!.findViewById(R.id.youtube_player_view)
+    }
+
+    @Test
+    fun testSingleClickListener() {
+        val mockClickListener = mock(View.OnClickListener::class.java)
+
+        youTubePlayerView!!.setOnClickListener(mockClickListener)
+        onView(withId(clickTestViewId)).perform(click())
+
+        verify(mockClickListener).onClick(youTubePlayerView)
+    }
+
+    @Test
+    fun testLongClickListener() {
+        val mockLongClickListener = mock(View.OnLongClickListener::class.java)
+
+        youTubePlayerView!!.setOnLongClickListener(mockLongClickListener)
+        onView(withId(clickTestViewId)).perform(longClick())
+
+        verify(mockLongClickListener).onLongClick(youTubePlayerView)
+    }
+
+    @Test
+    fun testTouchSwipeDownListener() {
+        val touchEventQueue: Queue<MotionEvent> = LinkedList()
+
+        youTubePlayerView!!.setOnTouchListener { view, motionEvent ->
+            touchEventQueue.add(
+                    MotionEvent.obtain(motionEvent)
+            )
+            false
+        }
+
+        onView(withId(clickTestViewId)).perform(swipeDown())
+
+        var prevMotionEvent: MotionEvent? = null
+        while (touchEventQueue.isNotEmpty()) {
+            val actualEventAction = touchEventQueue.poll()
+            if (actualEventAction!!.action == MotionEvent.ACTION_DOWN) {
+                prevMotionEvent = actualEventAction
+            } else if (actualEventAction.action == MotionEvent.ACTION_MOVE) {
+                assertEquals(prevMotionEvent!!.x, actualEventAction.x)
+                assertTrue(prevMotionEvent.y < actualEventAction.y)
+
+                prevMotionEvent = actualEventAction
+            }
+        }
+        assertNotNull(prevMotionEvent)
```

Comment:
if queue is empty (touch fails) then `prevMotionEvent` will remain null

Reply:
It would be better to find a more explicit way of testing that.

Maybe set a boolean to true once the touch happens and then assert that?
Otherwise mockito has a `spy` method that you can use to create mocks and maintain your implementation logic. I think you could: create the listener, create a spy object of it and then set the spy as the listener. Like:

```
val listener = TouchListener { // impl }
val spyListener = spy(listener)
view.addListener(spyListener)
verify(spyListener).onTouch()
```

This should be possible, correct me if I'm wrong though.

Reply:
I tried a lot of ways to make the `spy()` way work. But it doesn't seem to work on listeners. So, i'm using the boolean way instead